iis/mymodule.cpp

```diff
@@ -752,6 +762,58 @@ CMyHttpModule::OnBeginRequest(
         goto Finished;
 	}
 
+    // Get request body without holding lock, because some clients may be slow at sending
+    LeaveCriticalSection(&m_csLock);
+    HTTP_REQUEST * pRawRequest = pRequest->GetRawHttpRequest();
+    preAllocBodyChunk* head = NULL;
+    preAllocBodyChunk* cur = NULL;
+    while (pRequest->GetRemainingEntityBytes() > 0)
+    {
+        // Allocate memory for the preAllocBodyChunk linked list structure, and also the actual body content
+        // HUGE_STRING_LEN is hardcoded because this is also hardcoded in apache2_io.c's call to ap_get_brigade
+        preAllocBodyChunk* chunk = (preAllocBodyChunk*)malloc(sizeof(preAllocBodyChunk) + HUGE_STRING_LEN);
+        chunk->next = NULL;
+
+        // Pointer to rest of allocated memory, for convenience
+        chunk->data = chunk + 1;
+
+        DWORD readcnt;
+        HRESULT hr = pRequest->ReadEntityBody(chunk->data, HUGE_STRING_LEN, false, &readcnt, NULL);
+        if (ERROR_HANDLE_EOF == (hr & 0x0000FFFF))
+        {
+            free(chunk);
+            break;
+        }
+        chunk->length = readcnt;
+
+        if (!head)
+        {
+            // Initialize linked list
+            head = chunk;
+            cur = chunk;
+        }
+        else {
+            // Append to linked list
+            cur->next = chunk;
+            cur = chunk;
+        }
+
+        if (hr != S_OK)
+        {
+            goto Finished;
```

Comment:
In this code path LeaveCriticalSection would be called twice. You would need to EnterCriticalSection before goto Finished.

Reply:
another option will be to add another goto right after the release of the lock

Reply:
refactored into a func without the goto, so now always re-entering crit sec

Reply:
why not save the time of the lock/unlock if it is not needed?

Reply:
ok, added extra goto label